lib/tasks/chf_tasks.rake

```diff
@@ -214,6 +214,169 @@ namespace :chf do
+        begin
+          # A bit expensive to get all the id and checksums, is there a faster way? Not sure.
+          file = fs.original_file
+          if file
+            CHF::CreateDziService.new(file.id, checksum: file.checksum.value).call(lazy: lazy)
+          else
+            Rails.logger.warn("No original file for #{fs.id}? Could not push DZI")
+          end
+          progress.increment
+        rescue StandardError => e
+          errors << file.id
+          progress.log("Could not create and push DZI for #{file.id}: #{e}")
+        end
+      end
+      progress.finish
+      if errors.count
+        $stderr.puts "#{errors.count} errors"
+      end
+    end
+
+    # not sure why these 'require' are required
+    require 'hydra/pcdm'
+    desc "remove .dzi and files not associated with current fedora data"
+    task :clean_orphaned => :environment do |t, args|
+      # get a list of ALL top-level objects, which will be just .dzi files.
+      bucket = CHF::CreateDziService.s3_bucket!
+      scope = bucket.objects(delimiter: '/')
+
+      $stderr.puts "Scanning S3 bucket '#{bucket.name}' for .dzi of file IDs not currently in repo '#{ActiveFedora.fedora.base_uri}'...\n\n"
+
+      # Don't know total, too expensive to look up.
+      progress = ProgressBar.create(:total => nil)
+      i = 0
+      deleted = []
+
+      objects = scope.each do |s3_obj|
+        i += 1
+
+        file_id, checksum = CHF::CreateDziService.parse_dzi_file_name( s3_obj.key )
+
+        # Believe it or not, this seems to be the way to figure out file existence, took
+        # me hours to figure out. Not sure how many round-trips to fedora it will
+        # take, but this is good enough, i'm tired.
+        file_obj = Hydra::PCDM::File.new(file_id)
+        exists  = begin
+                    file_obj.persisted?
+                  rescue Ldp::Gone
+                    false
+                  end
+        stored_checksum = begin
+                            file_obj.checksum.value
+                          rescue Ldp::Gone
+                            false
+                          end
+
+        unless exists && stored_checksum == checksum
+          # either file_id does not exist, or no longer has this checksum.
+          # orphaned!
+          deleted << [file_id, checksum]
+
+          #first .dzi, so it won't be visible to front-end
+          s3_obj.delete
+
+          # then any associated tiles
+          bucket.objects(prefix: s3_obj.key.sub(/\.dzi$/, '_files/')).each do |tile_obj|
+            progress.increment
+            tile_obj.delete
+          end
+        end
+        progress.increment
+        progress.title = "#{i} scanned, #{deleted.count} deleted"
+      end
+      progress.finish
+
+      # Have to do this a bit hackily, we'll actually iterate through every
+      # key, but the sdk #list_objects methods gets 'directories' out
+      # for us with #prefix
+      $stderr.puts "\nScanning for orphaned _files/ tiles...\n\n"
```

Comment:
Interesting, have you observed this type of orphaned file? Any idea what would cause this?